File: solutions/125.valid-palindrome/src/two_pointers.ts
/*
 * @lc app=leetcode.cn id=125 lang=typescript
 *
 * https://leetcode.cn/problems/valid-palindrome/
 */

// @lc code=start
function isPalindrome(s: string): boolean {
    let chars = [...s]
        .filter((c) => c.match(/[a-zA-Z0-9]/))
        .map((c) => c.toLowerCase());
    if (chars.length < 2) {
        return true;
    }
    let i = 0;
    let j = chars.length - 1;
    while (i < j) {
        if (chars[i] !== chars[j]) {
            return false;
        }
        i++;
        j--;
    }
    return true;
}
// @lc code=end

export { isPalindrome };

File: solutions/125.valid-palindrome/src/two_pointers.rs
/*
 * @lc app=leetcode.cn id=125 lang=rust
 *
 * https://leetcode.cn/problems/valid-palindrome/
 */

pub struct Solution;

// @lc code=start
impl Solution {
    pub fn is_palindrome(s: String) -> bool {
        let s: Vec<char> = s
            .chars()
            .filter(|c| c.is_alphanumeric())
            .map(|c| c.to_ascii_lowercase())
            .collect();
        if s.len() < 2 {
            return true;
        }
        let mut i = 0;
        let mut j = s.len() - 1;
        while i < j {
            if s[i] != s[j] {
                return false;
            }
            i += 1;
            j -= 1;
        }
        true
    }
}
// @lc code=end
